<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Large-scale clustering in Bioconductor • cshlgsd2022</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Large-scale clustering in Bioconductor">
<meta property="og:description" content="Learn how to do unsupervised clustering in R/Bioconductor for large-scale data.
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cshlgsd2022</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/cshlgsd2022.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro-single-cell-01.html">Introduction to Single-Cell</a>
    </li>
    <li>
      <a href="../articles/intro-single-cell-02.html">Overview of single-cell analysis in R/Bioconductor</a>
    </li>
    <li>
      <a href="../articles/adv-single-cell-01.html">Large-scale clustering in Bioconductor</a>
    </li>
    <li>
      <a href="../articles/intro-spatial-01.html">Introduction to Spatial Transcriptomics Data Science</a>
    </li>
    <li>
      <a href="../articles/adv-spatial-01.html">Advanced Spatial Transcriptomics Data Science</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/index.html">More articles...</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://twitter.com/stephaniehicks" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/stephaniehicks/cshlgsd2022" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Large-scale clustering in Bioconductor</h1>
                        <h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            
      
      
      <div class="hidden name"><code>adv-single-cell-01.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="key-resources">Key resources<a class="anchor" aria-label="anchor" href="#key-resources"></a>
</h3>
<ul>
<li>Workshop material: <a href="https://stephaniehicks.com/cshlgsd2022">pkgdown website</a>
</li>
<li>Code: <a href="https://github.com/stephaniehicks/cshlgsd2022" class="external-link">GitHub</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-1">Part 1<a class="anchor" aria-label="anchor" href="#part-1"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Add here.</li>
<li>Add here.</li>
<li>Add here.</li>
<li>Add here.</li>
<li>Add here.</li>
</ol>
</div>
<div class="section level3">
<h3 id="materials">Materials<a class="anchor" aria-label="anchor" href="#materials"></a>
</h3>
<p>We will go through these slides on “More Single-Cell Data
Science”:</p>
<ul>
<li><a href="https://docs.google.com/presentation/d/19EIrGrbZoQUyYQggy8Zxzo5RiwM7fEsiP3DSAsnCOgE/edit?usp=sharing" class="external-link uri">https://docs.google.com/presentation/d/19EIrGrbZoQUyYQggy8Zxzo5RiwM7fEsiP3DSAsnCOgE/edit?usp=sharing</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-2">Part 2<a class="anchor" aria-label="anchor" href="#part-2"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives-1">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-1"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Be able to load a <code>SingleCellExperiment</code> object in a
<code>.HDF5</code> file format for large-scale single-cell data.</li>
<li>Be able to apply standard scRNA-seq workflow, but to a large
dataset.</li>
</ol>
</div>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>This tutorial gives a demo of large-scale clustering in the
Bioconductor using the <em>DelayedArray</em> framework.
<em>DelayedArray</em> is like an ordinary array in R, but allows for the
data to be in-memory, on-disk in a file, or even hosted on a remote
server.</p>
<p>We will showcase an end-to-end clustering pipeline, starting from the
count data matrix stored in HDF5 (similar to what one would download
from the HCA data portal) all the way to visualization and
interpretation of the clustering results.</p>
<ol style="list-style-type: decimal">
<li>
<code>scran</code> normalization + PCA</li>
<li>
<code>glmpca</code> with <code>scry</code>
</li>
</ol>
<p>Using these reduced dimensions, we present two types of
clustering:</p>
<ol style="list-style-type: decimal">
<li>SNN + Louvain clustering</li>
<li>mini-batch <em>k</em>-means (<code>mbkmeans</code>)</li>
</ol>
<p>While we use a small-ish dataset for this demo for convenience, the
code is computationally efficient even for (very) large datasets.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxPBMCData</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/scry.html" class="external-link">scry</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mbkmeans</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="getting-the-data">Getting the data<a class="anchor" aria-label="anchor" href="#getting-the-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html" class="external-link">TENxPBMCData</a></span><span class="op">(</span><span class="st">"pbmc4k"</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 33694 4340 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span></span>
<span><span class="co">##   ENSG00000268674</span></span>
<span><span class="co">## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(11): Sample Barcode ... Individual Date_published</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## &lt;33694 x 4340&gt; sparse matrix of class DelayedMatrix and type "integer":</span></span>
<span><span class="co">##                    [,1]    [,2]    [,3]    [,4] ... [,4337] [,4338] [,4339]</span></span>
<span><span class="co">## ENSG00000243485       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000237613       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000186092       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000238009       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000239945       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">##             ...       .       .       .       .   .       .       .       .</span></span>
<span><span class="co">## ENSG00000277856       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000275063       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000271254       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000277475       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">## ENSG00000268674       0       0       0       0   .       0       0       0</span></span>
<span><span class="co">##                 [,4340]</span></span>
<span><span class="co">## ENSG00000243485       0</span></span>
<span><span class="co">## ENSG00000237613       0</span></span>
<span><span class="co">## ENSG00000186092       0</span></span>
<span><span class="co">## ENSG00000238009       0</span></span>
<span><span class="co">## ENSG00000239945       0</span></span>
<span><span class="co">##             ...       .</span></span>
<span><span class="co">## ENSG00000277856       0</span></span>
<span><span class="co">## ENSG00000275063       0</span></span>
<span><span class="co">## ENSG00000271254       0</span></span>
<span><span class="co">## ENSG00000277475       0</span></span>
<span><span class="co">## ENSG00000268674       0</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seed</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "HDF5ArraySeed"</span></span>
<span><span class="co">## Slot "filepath":</span></span>
<span><span class="co">## [1] "/Users/stephaniehicks/Library/Caches/org.R-project.R/R/ExperimentHub/13b087bcbb296_1611"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "name":</span></span>
<span><span class="co">## [1] "/counts"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "as_sparse":</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "type":</span></span>
<span><span class="co">## [1] NA</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "dim":</span></span>
<span><span class="co">## [1] 33694  4340</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "chunkdim":</span></span>
<span><span class="co">## [1] 512  66</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "first_val":</span></span>
<span><span class="co">## [1] 0</span></span></code></pre>
<p>In this tutorial, we use a small dataset for the sake of running all
the code in a short amount of time. However, this workflow is designed
for large data and it will run just fine with any sized dataset. For
instance, we have analyzed 1.3 million cells on a machine with a
moderately sized RAM (e.g., 64GB).</p>
<p>By running the code below, you will run the workflow on the [10X
Genomics 1.3 million cells dataset. (Warning: it takes some time!)
Alternatively, you can substitute the code below with your own data in
<code>SingleCellExperiment</code> format.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxBrainData</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxBrainData/man/TENxBrainData.html" class="external-link">TENxBrainData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filtering-and-normalization">Filtering and normalization<a class="anchor" aria-label="anchor" href="#filtering-and-normalization"></a>
</h3>
<div class="section level4">
<h4 id="removing-low-quality-cells">Removing low-quality cells<a class="anchor" aria-label="anchor" href="#removing-low-quality-cells"></a>
</h4>
<p>First, we use the <code>scater</code> package to compute a set of QC
measures and filter out the low-quality samples.</p>
<p>Here, we exclude those cells that have a too high percentage of
mitochondrial genes or for which we detect too few genes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>            subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol_TENx</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">high_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">subsets_Mito_percent</span>, </span>
<span>                       nmads <span class="op">=</span> <span class="fl">3</span>, type<span class="op">=</span><span class="st">"higher"</span><span class="op">)</span></span>
<span><span class="va">low_detection</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">high_counts</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">sum</span> <span class="op">&gt;</span> <span class="fl">45000</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">high_mito</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">low_detection</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">high_counts</span><span class="op">]</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 33694 3401 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span></span>
<span><span class="co">##   ENSG00000268674</span></span>
<span><span class="co">## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(17): Sample Barcode ... subsets_Mito_percent total</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="removing-lowly-expressed-genes">Removing lowly expressed genes<a class="anchor" aria-label="anchor" href="#removing-lowly-expressed-genes"></a>
</h4>
<p>Next, we remove the lowly expressed genes. Here, we keep only those
genes that have at least 1 UMI in at least 5% of the data. These
threshold are dataset-specific and may need to be taylored to specific
applications.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_reads</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">num_cells</span> <span class="op">&lt;-</span> <span class="fl">0.01</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">DelayedArray</span><span class="fu">::</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">num_reads</span> <span class="op">)</span> <span class="op">&gt;=</span> <span class="va">num_cells</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 10944 3401 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(10944): ENSG00000279457 ENSG00000228463 ... ENSG00000273748</span></span>
<span><span class="co">##   ENSG00000278817</span></span>
<span><span class="co">## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(17): Sample Barcode ... subsets_Mito_percent total</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>These leaves us with <code>length(keep)</code> genes.</p>
</div>
<div class="section level4">
<h4 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h4>
<p>Here, we apply <code>mbkmeans</code> (<code>k=10</code> and batch
size of 500) as a preliminary step to <code>scran</code>
normalization.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">19</span><span class="op">)</span></span>
<span><span class="va">mbk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mbkmeans/man/mbkmeans.html" class="external-link">mbkmeans</a></span><span class="op">(</span><span class="va">sce</span>, whichAssay <span class="op">=</span> <span class="st">"counts"</span>, reduceMethod <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                  clusters<span class="op">=</span><span class="fl">10</span>, batch_size <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">mbk10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"mbk"</span>, <span class="va">mbk</span><span class="op">$</span><span class="va">Clusters</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">mbk</span><span class="op">$</span><span class="va">Clusters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##   1   2   3   4   5   6   7   8   9  10 </span></span>
<span><span class="co">##  88 405  64 259 433 766  86 594 700   6</span></span></code></pre>
<p>We then compute the normalization factors and normalize the data.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sce</span>, cluster<span class="op">=</span><span class="va">mbk</span><span class="op">$</span><span class="va">Clusters</span>, min.mean <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 10944 3401 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(10944): ENSG00000279457 ENSG00000228463 ... ENSG00000273748</span></span>
<span><span class="co">##   ENSG00000278817</span></span>
<span><span class="co">## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(19): Sample Barcode ... mbk10 sizeFactor</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<div class="section level4">
<h4 id="pca-on-normalized-values">PCA on normalized values<a class="anchor" aria-label="anchor" href="#pca-on-normalized-values"></a>
</h4>
<p>Here, we compute the first 50 principal components using the top
variable genes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                      ntop <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                      scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      BSPARAM <span class="op">=</span> <span class="fu">BiocSingular</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">RandomParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"mbk10"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/pca-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="glm-pca">GLM-PCA<a class="anchor" aria-label="anchor" href="#glm-pca"></a>
</h4>
<p>An alternative to PCA on normalized data is to use the GLM-PCA
approach, implemented in the <code>scry</code> Bioconductor package.
Here, we use the faster, approximate approach that computes the null
residuals and runs PCA on them.</p>
<p>Other approaches implemented in Bioconductor for dimensionality
reduction include correspondence analysis (in the <code>corral</code>
package) and ZINB-WaVE (in the <code>zinbwave</code> and
<code>NewWave</code> packages).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scry/man/nullResiduals.html" class="external-link">nullResiduals</a></span><span class="op">(</span><span class="va">sce</span>, assay<span class="op">=</span><span class="st">"counts"</span>, type<span class="op">=</span><span class="st">"deviance"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                      ntop <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                      exprs_values <span class="op">=</span> <span class="st">"binomial_deviance_residuals"</span>,</span>
<span>                      scale <span class="op">=</span> <span class="cn">TRUE</span>, name <span class="op">=</span> <span class="st">"GLM-PCA"</span>,</span>
<span>                      BSPARAM <span class="op">=</span> <span class="fu">BiocSingular</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">RandomParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"GLM-PCA"</span>, colour_by <span class="op">=</span> <span class="st">"mbk10"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/glmpca-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>Here, we use the GLM-PCA results to obtain the final cluster labels.
We use two alternative approaches: Louvain and mini-batch
<em>k</em>-means.</p>
<div class="section level4">
<h4 id="louvain">Louvain<a class="anchor" aria-label="anchor" href="#louvain"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">sce</span>, k<span class="op">=</span><span class="fl">10</span>, use.dimred <span class="op">=</span> <span class="st">"GLM-PCA"</span><span class="op">)</span></span>
<span><span class="va">lou</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_louvain.html" class="external-link">cluster_louvain</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">louvain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Louvain"</span>, <span class="va">lou</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">louvain</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Louvain1 Louvain2 Louvain3 Louvain4 Louvain5 Louvain6 </span></span>
<span><span class="co">##      871     1057      442      337      654       40</span></span></code></pre>
<p>If you want more control on the resolution of the clustering, you can
use the Louvain implementation available in the <code>resolution</code>
package. Alternatively, the <code>leiden</code> package implements the
Leiden algorithm.</p>
</div>
<div class="section level4">
<h4 id="mini-batch-k-means">Mini-batch k-means<a class="anchor" aria-label="anchor" href="#mini-batch-k-means"></a>
</h4>
<p>Mini-batch <span class="math inline">\(k\)</span>-means is a faster
version of <span class="math inline">\(k\)</span>-means that uses only a
random “mini-batch” of data at each iteration. The algorithm is fast
enough to cluster the 1.3 million cell data in the space of the top 50
PC in under 30 seconds.</p>
<p>Here, we run it multiple times to select the value of <span class="math inline">\(k\)</span> with the elbow method.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">km_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">k_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/mbkmeans/man/mbkmeans.html" class="external-link">mbkmeans</a></span><span class="op">(</span><span class="va">sce</span>, clusters <span class="op">=</span> <span class="va">k</span>, </span>
<span>             batch_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>             reduceMethod <span class="op">=</span> <span class="st">"GLM-PCA"</span>,</span>
<span>             calc_wcss <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">wcss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">km_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">WCSS_per_cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">k_list</span>, <span class="va">wcss</span>, type <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/select_k-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">kmeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"mbk"</span>, <span class="va">km_res</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">k_list</span><span class="op">==</span><span class="fl">12</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Clusters</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">kmeans</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  mbk1 mbk10 mbk11 mbk12  mbk2  mbk3  mbk4  mbk5  mbk6  mbk7  mbk8  mbk9 </span></span>
<span><span class="co">##   277    93    75     1   427    47   598    41   350     1   434  1057</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">kmeans</span>, <span class="va">sce</span><span class="op">$</span><span class="va">louvain</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        </span></span>
<span><span class="co">##         Louvain1 Louvain2 Louvain3 Louvain4 Louvain5 Louvain6</span></span>
<span><span class="co">##   mbk1       277        0        0        0        0        0</span></span>
<span><span class="co">##   mbk10       93        0        0        0        0        0</span></span>
<span><span class="co">##   mbk11       74        0        0        0        0        1</span></span>
<span><span class="co">##   mbk12        0        1        0        0        0        0</span></span>
<span><span class="co">##   mbk2       425        0        1        0        0        1</span></span>
<span><span class="co">##   mbk3         0        0        0       44        3        0</span></span>
<span><span class="co">##   mbk4         0       13        0        2      583        0</span></span>
<span><span class="co">##   mbk5         1        0        2        0        0       38</span></span>
<span><span class="co">##   mbk6         1       35        8      287       19        0</span></span>
<span><span class="co">##   mbk7         0        0        1        0        0        0</span></span>
<span><span class="co">##   mbk8         0        3      430        0        1        0</span></span>
<span><span class="co">##   mbk9         0     1005        0        4       48        0</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="cluster-visualization">Cluster visualization<a class="anchor" aria-label="anchor" href="#cluster-visualization"></a>
</h3>
<p>We can use UMAP or t-SNE to visualize the clusters.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"GLM-PCA"</span>, </span>
<span>                       external_neighbors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       BNPARAM <span class="op">=</span> <span class="fu">BiocNeighbors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/umap-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"kmeans"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/umap-2.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"GLM-PCA"</span>, </span>
<span>                       external_neighbors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       BNPARAM <span class="op">=</span> <span class="fu">BiocNeighbors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/tsne-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"kmeans"</span><span class="op">)</span></span></code></pre></div>
<p><img src="adv-single-cell-01_files/figure-html/tsne-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin21.5.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /opt/homebrew/Cellar/openblas/0.3.20/lib/libopenblasp-r0.3.20.dylib</span></span>
<span><span class="co">## LAPACK: /opt/homebrew/Cellar/r/4.2.1/lib/R/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] mbkmeans_1.12.0             scry_1.8.0                 </span></span>
<span><span class="co">##  [3] scran_1.24.0                scater_1.24.0              </span></span>
<span><span class="co">##  [5] ggplot2_3.3.6               scuttle_1.6.2              </span></span>
<span><span class="co">##  [7] TENxPBMCData_1.14.0         HDF5Array_1.24.1           </span></span>
<span><span class="co">##  [9] rhdf5_2.40.0                DelayedArray_0.22.0        </span></span>
<span><span class="co">## [11] Matrix_1.4-1                SingleCellExperiment_1.18.0</span></span>
<span><span class="co">## [13] SummarizedExperiment_1.26.1 Biobase_2.56.0             </span></span>
<span><span class="co">## [15] GenomicRanges_1.48.0        GenomeInfoDb_1.32.2        </span></span>
<span><span class="co">## [17] IRanges_2.30.0              S4Vectors_0.34.0           </span></span>
<span><span class="co">## [19] BiocGenerics_0.42.0         MatrixGenerics_1.8.1       </span></span>
<span><span class="co">## [21] matrixStats_0.62.0         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] AnnotationHub_3.4.0           BiocFileCache_2.4.0          </span></span>
<span><span class="co">##   [3] systemfonts_1.0.4             igraph_1.3.2                 </span></span>
<span><span class="co">##   [5] gmp_0.6-5                     BiocParallel_1.30.3          </span></span>
<span><span class="co">##   [7] benchmarkme_1.0.8             digest_0.6.29                </span></span>
<span><span class="co">##   [9] foreach_1.5.2                 htmltools_0.5.2              </span></span>
<span><span class="co">##  [11] viridis_0.6.2                 fansi_1.0.3                  </span></span>
<span><span class="co">##  [13] magrittr_2.0.3                memoise_2.0.1                </span></span>
<span><span class="co">##  [15] ScaledMatrix_1.4.0            doParallel_1.0.17            </span></span>
<span><span class="co">##  [17] cluster_2.1.3                 limma_3.52.2                 </span></span>
<span><span class="co">##  [19] Biostrings_2.64.0             pkgdown_2.0.5                </span></span>
<span><span class="co">##  [21] colorspace_2.0-3              blob_1.2.3                   </span></span>
<span><span class="co">##  [23] rappdirs_0.3.3                ggrepel_0.9.1                </span></span>
<span><span class="co">##  [25] textshaping_0.3.6             xfun_0.31                    </span></span>
<span><span class="co">##  [27] dplyr_1.0.9                   crayon_1.5.1                 </span></span>
<span><span class="co">##  [29] RCurl_1.98-1.7                jsonlite_1.8.0               </span></span>
<span><span class="co">##  [31] iterators_1.0.14              glue_1.6.2                   </span></span>
<span><span class="co">##  [33] gtable_0.3.0                  zlibbioc_1.42.0              </span></span>
<span><span class="co">##  [35] XVector_0.36.0                BiocSingular_1.12.0          </span></span>
<span><span class="co">##  [37] Rhdf5lib_1.18.2               scales_1.2.0                 </span></span>
<span><span class="co">##  [39] DBI_1.1.3                     edgeR_3.38.1                 </span></span>
<span><span class="co">##  [41] Rcpp_1.0.8.3                  viridisLite_0.4.0            </span></span>
<span><span class="co">##  [43] xtable_1.8-4                  dqrng_0.3.0                  </span></span>
<span><span class="co">##  [45] bit_4.0.4                     rsvd_1.0.5                   </span></span>
<span><span class="co">##  [47] metapod_1.4.0                 httr_1.4.3                   </span></span>
<span><span class="co">##  [49] ellipsis_0.3.2                ClusterR_1.2.6               </span></span>
<span><span class="co">##  [51] farver_2.1.0                  pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [53] uwot_0.1.11                   sass_0.4.1                   </span></span>
<span><span class="co">##  [55] dbplyr_2.2.1                  locfit_1.5-9.5               </span></span>
<span><span class="co">##  [57] utf8_1.2.2                    labeling_0.4.2               </span></span>
<span><span class="co">##  [59] tidyselect_1.1.2              rlang_1.0.3                  </span></span>
<span><span class="co">##  [61] later_1.3.0                   AnnotationDbi_1.58.0         </span></span>
<span><span class="co">##  [63] munsell_0.5.0                 BiocVersion_3.15.2           </span></span>
<span><span class="co">##  [65] tools_4.2.1                   cachem_1.0.6                 </span></span>
<span><span class="co">##  [67] cli_3.3.0                     generics_0.1.2               </span></span>
<span><span class="co">##  [69] RSQLite_2.2.14                ExperimentHub_2.4.0          </span></span>
<span><span class="co">##  [71] evaluate_0.15                 stringr_1.4.0                </span></span>
<span><span class="co">##  [73] fastmap_1.1.0                 yaml_2.3.5                   </span></span>
<span><span class="co">##  [75] ragg_1.2.2                    knitr_1.39                   </span></span>
<span><span class="co">##  [77] bit64_4.0.5                   fs_1.5.2                     </span></span>
<span><span class="co">##  [79] purrr_0.3.4                   KEGGREST_1.36.2              </span></span>
<span><span class="co">##  [81] sparseMatrixStats_1.8.0       mime_0.12                    </span></span>
<span><span class="co">##  [83] compiler_4.2.1                rstudioapi_0.13              </span></span>
<span><span class="co">##  [85] beeswarm_0.4.0                filelock_1.0.2               </span></span>
<span><span class="co">##  [87] curl_4.3.2                    png_0.1-7                    </span></span>
<span><span class="co">##  [89] interactiveDisplayBase_1.34.0 tibble_3.1.7                 </span></span>
<span><span class="co">##  [91] statmod_1.4.36                bslib_0.3.1                  </span></span>
<span><span class="co">##  [93] stringi_1.7.6                 highr_0.9                    </span></span>
<span><span class="co">##  [95] RSpectra_0.16-1               desc_1.4.1                   </span></span>
<span><span class="co">##  [97] lattice_0.20-45               bluster_1.6.0                </span></span>
<span><span class="co">##  [99] vctrs_0.4.1                   pillar_1.7.0                 </span></span>
<span><span class="co">## [101] lifecycle_1.0.1               rhdf5filters_1.8.0           </span></span>
<span><span class="co">## [103] BiocManager_1.30.18           jquerylib_0.1.4              </span></span>
<span><span class="co">## [105] BiocNeighbors_1.14.0          bitops_1.0-7                 </span></span>
<span><span class="co">## [107] irlba_2.3.5                   httpuv_1.6.5                 </span></span>
<span><span class="co">## [109] R6_2.5.1                      promises_1.2.0.1             </span></span>
<span><span class="co">## [111] gridExtra_2.3                 vipor_0.4.5                  </span></span>
<span><span class="co">## [113] codetools_0.2-18              benchmarkmeData_1.0.4        </span></span>
<span><span class="co">## [115] gtools_3.9.2.2                assertthat_0.2.1             </span></span>
<span><span class="co">## [117] rprojroot_2.0.3               withr_2.5.0                  </span></span>
<span><span class="co">## [119] GenomeInfoDbData_1.2.8        parallel_4.2.1               </span></span>
<span><span class="co">## [121] grid_4.2.1                    beachmat_2.12.0              </span></span>
<span><span class="co">## [123] rmarkdown_2.14                DelayedMatrixStats_1.18.0    </span></span>
<span><span class="co">## [125] Rtsne_0.16                    shiny_1.7.1                  </span></span>
<span><span class="co">## [127] ggbeeswarm_0.6.0</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephanie C. Hicks.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
