<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Single-Cell • cshlgsd2022</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Single-Cell">
<meta property="og:description" content="Introduction to Single-Cell Data Science
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cshlgsd2022</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/cshlgsd2022.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro-single-cell-01.html">Introduction to Single-Cell</a>
    </li>
    <li>
      <a href="../articles/intro-single-cell-02.html">Overview of single-cell analysis in R/Bioconductor</a>
    </li>
    <li>
      <a href="../articles/adv-single-cell-01.html">Large-scale clustering in Bioconductor</a>
    </li>
    <li>
      <a href="../articles/intro-spatial-01.html">Introduction to Spatial Transcriptomics Data Science</a>
    </li>
    <li>
      <a href="../articles/adv-spatial-01.html">Advanced Spatial Transcriptomics Data Science</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/index.html">More articles...</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://twitter.com/stephaniehicks" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/stephaniehicks/cshlgsd2022" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Single-Cell</h1>
                        <h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            
      
      
      <div class="hidden name"><code>intro-single-cell-01.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="key-resources">Key resources<a class="anchor" aria-label="anchor" href="#key-resources"></a>
</h3>
<ul>
<li>Workshop material: <a href="https://stephaniehicks.com/cshlgsd2022">pkgdown website</a>
</li>
<li>Code: <a href="https://github.com/stephaniehicks/cshlgsd2022" class="external-link">GitHub</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-1">Part 1<a class="anchor" aria-label="anchor" href="#part-1"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Understand how count matrices are created from single-cell
experimental platforms and protocols</li>
<li>Recognize which basic principles and concepts were transfered from
bulk to single-cell data analyses</li>
<li>Understand the key differences between bulk and single-cell
data</li>
<li>Define what is a “Unique Molecular Identifier”</li>
<li>Define multiplexing (and demultiplexing)</li>
</ol>
</div>
<div class="section level3">
<h3 id="materials">Materials<a class="anchor" aria-label="anchor" href="#materials"></a>
</h3>
<p>We will go through the introductory single-cell data science slides
available here:</p>
<ul>
<li><a href="https://docs.google.com/presentation/d/19tLNkFzNP2zkWGdMeaTAdE3quv78ukUeH_-v85g_pDI/edit?usp=sharing" class="external-link uri">https://docs.google.com/presentation/d/19tLNkFzNP2zkWGdMeaTAdE3quv78ukUeH_-v85g_pDI/edit?usp=sharing</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-2">Part 2<a class="anchor" aria-label="anchor" href="#part-2"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives-1">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-1"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Be able to create a count matrix and read it into R</li>
<li>Recognize and define the <code>SingleCellExperiment</code> S4 class
in R/Bioconductor to store single-cell data</li>
<li>Understand strategies to get access to existing single-cell data in
R</li>
</ol>
</div>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>NGS data from scRNA-seq experiments must be converted into a matrix
of expression values. This is usually a count matrix containing the
number of reads (or UMIs) mapped to each gene (row) in each cell
(column). Once this quantification is complete, we can proceed with our
downstream statistical analyses in R.</p>
<p>Constructing a count matrix from raw scRNA-seq data requires some
thought as the term “single-cell RNA-seq” encompasses a variety of
different experimental protocols. This includes</p>
<ul>
<li>droplet-based protocols like 10X Genomics, inDrop and Drop-seq</li>
<li>plate-based protocols with UMIs like CEL-seq(2) and MARS-seq</li>
<li>plate-based protocols with reads (mostly Smart-seq2)</li>
<li>others like sciRNA-seq, etc</li>
</ul>
<p>Each approach requires a different processing pipeline to deal with
cell demultiplexing and UMI deduplication (if applicable). Here, we will
briefly describe some of the methods used to generate a count matrix and
read it into R.</p>
</div>
<div class="section level3">
<h3 id="creating-a-count-matrix">Creating a count matrix<a class="anchor" aria-label="anchor" href="#creating-a-count-matrix"></a>
</h3>
<p>As mentioned above, the exact procedure for quantifying expression
depends on the technology involved:</p>
<ul>
<li>For 10X Genomics data, the <code>Cellranger</code> software suite
(Zheng et al. 2017) provides a custom pipeline to obtain a count matrix.
This uses STAR to align reads to the reference genome and then counts
the number of unique UMIs mapped to each gene.</li>
<li>Alternatively, pseudo-alignment methods such as <a href="https://doi.org/10.1186/s13059-019-1670-y" class="external-link"><code>alevin</code></a>
(Srivastava et al. 2019) can be used to obtain a count matrix from the
same data. This avoids the need for explicit alignment, which reduces
the compute time and memory usage.</li>
<li>For other highly multiplexed protocols, the <a href="http://bioconductor.org/packages/scPipe/" class="external-link"><code>scPipe</code></a>
package provides a more general pipeline for processing scRNA-seq data.
This uses the Rsubread aligner to align reads and then counts reads or
UMIs per gene.</li>
<li>For CEL-seq or CEL-seq2 data, the <code>scruff</code> package
provides a dedicated pipeline for quantification.</li>
<li>For read-based protocols, we can generally re-use the same pipelines
for processing bulk RNA-seq data (e.g. Subread, RSEM, salmon)</li>
<li>For any data involving spike-in transcripts, the spike-in sequences
should be included as part of the reference genome during alignment and
quantification.</li>
</ul>
<p>In all cases, the identity of the genes in the count matrix should be
defined with standard identifiers from Ensembl or Entrez. These provide
an unambiguous mapping between each row of the matrix and the
corresponding gene.</p>
<p>In contrast, a single gene symbol may be used by multiple loci, or
the mapping between symbols and genes may change over time, e.g., if the
gene is renamed. This makes it difficult to re-use the count matrix as
we cannot be confident in the meaning of the symbols. (Of course,
identifiers can be easily converted to gene symbols later on in the
analysis. This is the recommended approach as it allows us to document
how the conversion was performed and to backtrack to the stable
identifiers if the symbols are ambiguous.)</p>
</div>
<div class="section level3">
<h3 id="singlecellexperiment-class">
<code>SingleCellExperiment</code> Class<a class="anchor" aria-label="anchor" href="#singlecellexperiment-class"></a>
</h3>
<p>One of the main strengths of the Bioconductor project lies in the use
of a common data infrastructure that powers interoperability across
packages. Users should be able to analyze their data using functions
from different Bioconductor packages without the need to convert between
formats. To this end, the <code>SingleCellExperiment</code> class (from
the <code>SingleCellExperiment</code> package) serves as the common
currency for data exchange across 70+ single-cell-related Bioconductor
packages. This class implements a data structure that stores all aspects
of our single-cell data - gene-by-cell expression data, per-cell
metadata and per-gene annotation - and manipulate them in a synchronized
manner.</p>
<p><img src="figures/SingleCellExperiment.png" width="780px" style="display: block; margin: auto;"></p>
<p>Amezquita et al. 2019 (<a href="https://doi.org/10.1101/590562" class="external-link uri">https://doi.org/10.1101/590562</a>)</p>
<p>Each piece of (meta)data in the <code>SingleCellExperiment</code> is
represented by a separate “slot”. (This terminology comes from the S4
class system, but that’s not important right now.) If we imagine the
<code>SingleCellExperiment</code> object to be a cargo ship, the slots
can be thought of as individual cargo boxes with different contents,
e.g., certain slots expect numeric matrices whereas others may expect
data frames.</p>
<p>If you want to know more about the available slots, their expected
formats, and how we can interact with them, check out this <a href="https://bioconductor.org/books/3.15/OSCA.intro/the-singlecellexperiment-class.html" class="external-link">chapter</a>.</p>
<div class="section level4">
<h4 id="singlecellexperiment-example">
<code>SingleCellExperiment</code> Example<a class="anchor" aria-label="anchor" href="#singlecellexperiment-example"></a>
</h4>
<p>Let’s show you what a <code>SingleCellExperiment</code> (or
<code>sce</code> for short) looks like.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 20006 3005 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(20006): Tspan12 Tshz1 ... mt-Rnr1 mt-Nd4l</span></span>
<span><span class="co">## rowData names(1): featureType</span></span>
<span><span class="co">## colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12</span></span>
<span><span class="co">##   1772058148_F03</span></span>
<span><span class="co">## colData names(10): tissue group # ... level1class level2class</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: endogenous</span></span>
<span><span class="co">## altExpNames(2): ERCC repeat</span></span></code></pre>
<p>This <code>SingleCellExperiment</code> object has 20006 genes and
3005 cells.</p>
<p>We can pull out the counts matrix with the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts()</a></code>
function and the corresponding <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          1772071015_C02 1772071017_G12 1772071017_A05 1772071014_B06</span></span>
<span><span class="co">## Tspan12               0              0              0              3</span></span>
<span><span class="co">## Tshz1                 3              1              0              2</span></span>
<span><span class="co">## Fnbp1l                3              1              6              4</span></span>
<span><span class="co">## Adamts15              0              0              0              0</span></span>
<span><span class="co">## Cldn12                1              1              1              0</span></span>
<span><span class="co">##          1772067065_H06</span></span>
<span><span class="co">## Tspan12               0</span></span>
<span><span class="co">## Tshz1                 2</span></span>
<span><span class="co">## Fnbp1l                1</span></span>
<span><span class="co">## Adamts15              0</span></span>
<span><span class="co">## Cldn12                0</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20006 rows and 1 column</span></span>
<span><span class="co">##          featureType</span></span>
<span><span class="co">##          &lt;character&gt;</span></span>
<span><span class="co">## Tspan12   endogenous</span></span>
<span><span class="co">## Tshz1     endogenous</span></span>
<span><span class="co">## Fnbp1l    endogenous</span></span>
<span><span class="co">## Adamts15  endogenous</span></span>
<span><span class="co">## Cldn12    endogenous</span></span>
<span><span class="co">## ...              ...</span></span>
<span><span class="co">## mt-Co2          mito</span></span>
<span><span class="co">## mt-Co1          mito</span></span>
<span><span class="co">## mt-Rnr2         mito</span></span>
<span><span class="co">## mt-Rnr1         mito</span></span>
<span><span class="co">## mt-Nd4l         mito</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3005 rows and 10 columns</span></span>
<span><span class="co">##                        tissue   group # total mRNA mol      well       sex</span></span>
<span><span class="co">##                   &lt;character&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1772071015_C02       sscortex         1           1221         3         3</span></span>
<span><span class="co">## 1772071017_G12       sscortex         1           1231        95         1</span></span>
<span><span class="co">## 1772071017_A05       sscortex         1           1652        27         1</span></span>
<span><span class="co">## 1772071014_B06       sscortex         1           1696        37         3</span></span>
<span><span class="co">## 1772067065_H06       sscortex         1           1219        43         3</span></span>
<span><span class="co">## ...                       ...       ...            ...       ...       ...</span></span>
<span><span class="co">## 1772067059_B04 ca1hippocampus         9           1997        19         1</span></span>
<span><span class="co">## 1772066097_D04 ca1hippocampus         9           1415        21         1</span></span>
<span><span class="co">## 1772063068_D01       sscortex         9           1876        34         3</span></span>
<span><span class="co">## 1772066098_A12 ca1hippocampus         9           1546        88         1</span></span>
<span><span class="co">## 1772058148_F03       sscortex         9           1970        15         3</span></span>
<span><span class="co">##                      age  diameter        cell_id       level1class level2class</span></span>
<span><span class="co">##                &lt;numeric&gt; &lt;numeric&gt;    &lt;character&gt;       &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1772071015_C02         2         1 1772071015_C02      interneurons       Int10</span></span>
<span><span class="co">## 1772071017_G12         1       353 1772071017_G12      interneurons       Int10</span></span>
<span><span class="co">## 1772071017_A05         1        13 1772071017_A05      interneurons        Int6</span></span>
<span><span class="co">## 1772071014_B06         2        19 1772071014_B06      interneurons       Int10</span></span>
<span><span class="co">## 1772067065_H06         6        12 1772067065_H06      interneurons        Int9</span></span>
<span><span class="co">## ...                  ...       ...            ...               ...         ...</span></span>
<span><span class="co">## 1772067059_B04         4       382 1772067059_B04 endothelial-mural       Peric</span></span>
<span><span class="co">## 1772066097_D04         7        12 1772066097_D04 endothelial-mural        Vsmc</span></span>
<span><span class="co">## 1772063068_D01         7       268 1772063068_D01 endothelial-mural        Vsmc</span></span>
<span><span class="co">## 1772066098_A12         7       324 1772066098_A12 endothelial-mural        Vsmc</span></span>
<span><span class="co">## 1772058148_F03         7         6 1772058148_F03 endothelial-mural        Vsmc</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="data-resources">Data resources<a class="anchor" aria-label="anchor" href="#data-resources"></a>
</h3>
<p>In this section, we will discuss data packages and website for where
to get existing single-cell data</p>
<div class="section level4">
<h4 id="hcadata-data-package">
<code>HCAData</code> data package<a class="anchor" aria-label="anchor" href="#hcadata-data-package"></a>
</h4>
<p>The <code>HCAData</code> package allows a direct access to the
dataset generated by the Human Cell Atlas project for further processing
in R and Bioconductor. It does so by providing the datasets as
<code>SingleCellExperiment</code> objects. The datasets use
<code>HDF5Array</code> package to avoid loading the entire data set in
memory. Instead, it stores the counts on disk as a <code>.HDF5</code>
file, and loads subsets of the data into memory upon request.</p>
<p>The datasets are otherwise available in other formats (also as raw
data) at this link: <a href="http://preview.data.humancellatlas.org/" class="external-link uri">http://preview.data.humancellatlas.org/</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/federicomarini/HCAData" class="external-link">"HCAData"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HCAData/man/HCAData.html" class="external-link">HCAData</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/HCAData/man/HCAData.html" class="external-link">HCAData()</a></code> function downloads the relevant files from
<code>ExperimentHub</code>. If no argument is provided, a list of the
available datasets is returned, specifying which name to enter as
dataset parameter when calling HCAData.</p>
<p>If we specify either <code>ica_bone_marrow</code> or
<code>ica_cord_blood</code> in the function, we get returend a
<code>SingleCellExperiment</code> object</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_bonemarrow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HCAData/man/HCAData.html" class="external-link">HCAData</a></span><span class="op">(</span><span class="st">"ica_bone_marrow"</span><span class="op">)</span></span>
<span><span class="va">sce_bonemarrow</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 33694 378000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span></span>
<span><span class="co">##   ENSG00000268674</span></span>
<span><span class="co">## rowData names(2): ID Symbol</span></span>
<span><span class="co">## colnames(378000): MantonBM1_HiSeq_1-AAACCTGAGCAGGTCA-1</span></span>
<span><span class="co">##   MantonBM1_HiSeq_1-AAACCTGCACACTGCG-1 ...</span></span>
<span><span class="co">##   MantonBM8_HiSeq_8-TTTGTCATCTGCCAGG-1</span></span>
<span><span class="co">##   MantonBM8_HiSeq_8-TTTGTCATCTTGAGAC-1</span></span>
<span><span class="co">## colData names(1): Barcode</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>We can see even though it’s a lot cells, this is actually quite small
of an object in terms of data read into memory. This is due to the magic
of <code>HDF5Array</code> and <code>DelayedArray</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pryr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pryr/man/object_size.html" class="external-link">object_size</a></span><span class="op">(</span><span class="va">sce_bonemarrow</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 44.17 MB</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="hca-data-package">
<code>hca</code> data package<a class="anchor" aria-label="anchor" href="#hca-data-package"></a>
</h4>
<p>The <code>HCAData</code> package only has two datasets from the Human
Cell Atlas, but for a comprehensive list fromt he HCA Data Coordinating
Platform, check out the <code>hca</code> R/Bioconductor package (<a href="http://www.bioconductor.org/packages/hca" class="external-link uri">http://www.bioconductor.org/packages/hca</a>).</p>
</div>
<div class="section level4">
<h4 id="scrnaseq-data-package">
<code>scRNAseq</code> data package<a class="anchor" aria-label="anchor" href="#scrnaseq-data-package"></a>
</h4>
<p>The <code>scRNAseq</code> package provides convenient access to
several publicly available data sets in the form of
<code>SingleCellExperiment</code> objects. The focus of this package is
to capture datasets that are not easily read into R with a one-liner
from, e.g., read.csv(). Instead, we do the necessary data munging so
that users only need to call a single function to obtain a well-formed
<code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html" class="external-link">ZeiselBrainData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 20006 3005 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(20006): Tspan12 Tshz1 ... mt-Rnr1 mt-Nd4l</span></span>
<span><span class="co">## rowData names(1): featureType</span></span>
<span><span class="co">## colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12</span></span>
<span><span class="co">##   1772058148_F03</span></span>
<span><span class="co">## colData names(10): tissue group # ... level1class level2class</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: endogenous</span></span>
<span><span class="co">## altExpNames(2): ERCC repeat</span></span></code></pre>
<p>To see the list of available datasets, use the
<code><a href="https://rdrr.io/pkg/scRNAseq/man/listDatasets.html" class="external-link">listDatasets()</a></code> function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/listDatasets.html" class="external-link">listDatasets</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">out</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 61 rows and 5 columns</span></span>
<span><span class="co">##                  Reference  Taxonomy               Part    Number</span></span>
<span><span class="co">##                &lt;character&gt; &lt;integer&gt;        &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1   @aztekin2019identifi..      8355               tail     13199</span></span>
<span><span class="co">## 2   @bach2017differentia..     10090      mammary gland     25806</span></span>
<span><span class="co">## 3           @bacher2020low      9606            T cells    104417</span></span>
<span><span class="co">## 4     @baron2016singlecell      9606           pancreas      8569</span></span>
<span><span class="co">## 5     @baron2016singlecell     10090           pancreas      1886</span></span>
<span><span class="co">## ...                    ...       ...                ...       ...</span></span>
<span><span class="co">## 57    @zeisel2018molecular     10090     nervous system    160796</span></span>
<span><span class="co">## 58     @zhao2020singlecell      9606 liver immune cells     68100</span></span>
<span><span class="co">## 59    @zhong2018singlecell      9606  prefrontal cortex      2394</span></span>
<span><span class="co">## 60  @zilionis2019singlec..      9606               lung    173954</span></span>
<span><span class="co">## 61  @zilionis2019singlec..     10090               lung     17549</span></span>
<span><span class="co">##                       Call</span></span>
<span><span class="co">##                &lt;character&gt;</span></span>
<span><span class="co">## 1        AztekinTailData()</span></span>
<span><span class="co">## 2        BachMammaryData()</span></span>
<span><span class="co">## 3        BacherTCellData()</span></span>
<span><span class="co">## 4   BaronPancreasData('h..</span></span>
<span><span class="co">## 5   BaronPancreasData('m..</span></span>
<span><span class="co">## ...                    ...</span></span>
<span><span class="co">## 57     ZeiselNervousData()</span></span>
<span><span class="co">## 58   ZhaoImmuneLiverData()</span></span>
<span><span class="co">## 59   ZhongPrefrontalData()</span></span>
<span><span class="co">## 60      ZilionisLungData()</span></span>
<span><span class="co">## 61  ZilionisLungData('mo..</span></span></code></pre>
<p>If the original dataset was not provided with Ensembl annotation, we
can map the identifiers with <code>ensembl=TRUE</code>. Any genes
without a corresponding <code>Ensembl</code> identifier is discarded
from the dataset.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html" class="external-link">ZeiselBrainData</a></span><span class="op">(</span>ensembl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ENSMUSG00000029669" "ENSMUSG00000046982" "ENSMUSG00000039735"</span></span>
<span><span class="co">## [4] "ENSMUSG00000033453" "ENSMUSG00000046798" "ENSMUSG00000034009"</span></span></code></pre>
<p>Functions also have a <code>location=TRUE</code> argument that loads
in the gene coordinates.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html" class="external-link">ZeiselBrainData</a></span><span class="op">(</span>ensembl<span class="op">=</span><span class="cn">TRUE</span>, location<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">##                      seqnames              ranges strand | featureType</span></span>
<span><span class="co">##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   ENSMUSG00000029669        6   21771395-21852515      - |  endogenous</span></span>
<span><span class="co">##   ENSMUSG00000046982       18   84011627-84087706      - |  endogenous</span></span>
<span><span class="co">##   ENSMUSG00000039735        3 122538719-122619715      - |  endogenous</span></span>
<span><span class="co">##   ENSMUSG00000033453        9   30899155-30922452      - |  endogenous</span></span>
<span><span class="co">##   ENSMUSG00000046798        5     5489537-5514958      - |  endogenous</span></span>
<span><span class="co">##   ENSMUSG00000034009        3   79641611-79737880      - |  endogenous</span></span>
<span><span class="co">##                      originalName</span></span>
<span><span class="co">##                       &lt;character&gt;</span></span>
<span><span class="co">##   ENSMUSG00000029669      Tspan12</span></span>
<span><span class="co">##   ENSMUSG00000046982        Tshz1</span></span>
<span><span class="co">##   ENSMUSG00000039735       Fnbp1l</span></span>
<span><span class="co">##   ENSMUSG00000033453     Adamts15</span></span>
<span><span class="co">##   ENSMUSG00000046798       Cldn12</span></span>
<span><span class="co">##   ENSMUSG00000034009        Rxfp1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 118 sequences (1 circular) from GRCm38 genome</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin21.5.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /opt/homebrew/Cellar/openblas/0.3.20/lib/libopenblasp-r0.3.20.dylib</span></span>
<span><span class="co">## LAPACK: /opt/homebrew/Cellar/r/4.2.1/lib/R/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ensembldb_2.20.2            AnnotationFilter_1.20.0    </span></span>
<span><span class="co">##  [3] GenomicFeatures_1.48.3      AnnotationDbi_1.58.0       </span></span>
<span><span class="co">##  [5] rhdf5_2.40.0                HCAData_1.12.0             </span></span>
<span><span class="co">##  [7] scRNAseq_2.10.0             scater_1.24.0              </span></span>
<span><span class="co">##  [9] ggplot2_3.3.6               scuttle_1.6.2              </span></span>
<span><span class="co">## [11] SingleCellExperiment_1.18.0 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">## [13] Biobase_2.56.0              GenomicRanges_1.48.0       </span></span>
<span><span class="co">## [15] GenomeInfoDb_1.32.2         IRanges_2.30.0             </span></span>
<span><span class="co">## [17] S4Vectors_0.34.0            BiocGenerics_0.42.0        </span></span>
<span><span class="co">## [19] MatrixGenerics_1.8.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] AnnotationHub_3.4.0           BiocFileCache_2.4.0          </span></span>
<span><span class="co">##   [3] systemfonts_1.0.4             lazyeval_0.2.2               </span></span>
<span><span class="co">##   [5] BiocParallel_1.30.3           pryr_0.1.5                   </span></span>
<span><span class="co">##   [7] digest_0.6.29                 htmltools_0.5.2              </span></span>
<span><span class="co">##   [9] viridis_0.6.2                 fansi_1.0.3                  </span></span>
<span><span class="co">##  [11] magrittr_2.0.3                memoise_2.0.1                </span></span>
<span><span class="co">##  [13] ScaledMatrix_1.4.0            Biostrings_2.64.0            </span></span>
<span><span class="co">##  [15] pkgdown_2.0.5                 prettyunits_1.1.1            </span></span>
<span><span class="co">##  [17] colorspace_2.0-3              blob_1.2.3                   </span></span>
<span><span class="co">##  [19] rappdirs_0.3.3                ggrepel_0.9.1                </span></span>
<span><span class="co">##  [21] lobstr_1.1.2                  textshaping_0.3.6            </span></span>
<span><span class="co">##  [23] xfun_0.31                     dplyr_1.0.9                  </span></span>
<span><span class="co">##  [25] crayon_1.5.1                  RCurl_1.98-1.7               </span></span>
<span><span class="co">##  [27] jsonlite_1.8.0                glue_1.6.2                   </span></span>
<span><span class="co">##  [29] gtable_0.3.0                  zlibbioc_1.42.0              </span></span>
<span><span class="co">##  [31] XVector_0.36.0                DelayedArray_0.22.0          </span></span>
<span><span class="co">##  [33] BiocSingular_1.12.0           Rhdf5lib_1.18.2              </span></span>
<span><span class="co">##  [35] HDF5Array_1.24.1              scales_1.2.0                 </span></span>
<span><span class="co">##  [37] DBI_1.1.3                     Rcpp_1.0.8.3                 </span></span>
<span><span class="co">##  [39] viridisLite_0.4.0             xtable_1.8-4                 </span></span>
<span><span class="co">##  [41] progress_1.2.2                bit_4.0.4                    </span></span>
<span><span class="co">##  [43] rsvd_1.0.5                    httr_1.4.3                   </span></span>
<span><span class="co">##  [45] ellipsis_0.3.2                pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [47] XML_3.99-0.10                 sass_0.4.1                   </span></span>
<span><span class="co">##  [49] dbplyr_2.2.1                  utf8_1.2.2                   </span></span>
<span><span class="co">##  [51] tidyselect_1.1.2              rlang_1.0.3                  </span></span>
<span><span class="co">##  [53] later_1.3.0                   munsell_0.5.0                </span></span>
<span><span class="co">##  [55] BiocVersion_3.15.2            tools_4.2.1                  </span></span>
<span><span class="co">##  [57] cachem_1.0.6                  cli_3.3.0                    </span></span>
<span><span class="co">##  [59] generics_0.1.2                RSQLite_2.2.14               </span></span>
<span><span class="co">##  [61] ExperimentHub_2.4.0           evaluate_0.15                </span></span>
<span><span class="co">##  [63] stringr_1.4.0                 fastmap_1.1.0                </span></span>
<span><span class="co">##  [65] yaml_2.3.5                    ragg_1.2.2                   </span></span>
<span><span class="co">##  [67] knitr_1.39                    bit64_4.0.5                  </span></span>
<span><span class="co">##  [69] fs_1.5.2                      purrr_0.3.4                  </span></span>
<span><span class="co">##  [71] KEGGREST_1.36.2               sparseMatrixStats_1.8.0      </span></span>
<span><span class="co">##  [73] mime_0.12                     xml2_1.3.3                   </span></span>
<span><span class="co">##  [75] biomaRt_2.52.0                compiler_4.2.1               </span></span>
<span><span class="co">##  [77] rstudioapi_0.13               beeswarm_0.4.0               </span></span>
<span><span class="co">##  [79] filelock_1.0.2                curl_4.3.2                   </span></span>
<span><span class="co">##  [81] png_0.1-7                     interactiveDisplayBase_1.34.0</span></span>
<span><span class="co">##  [83] tibble_3.1.7                  bslib_0.3.1                  </span></span>
<span><span class="co">##  [85] stringi_1.7.6                 highr_0.9                    </span></span>
<span><span class="co">##  [87] desc_1.4.1                    lattice_0.20-45              </span></span>
<span><span class="co">##  [89] ProtGenerics_1.28.0           Matrix_1.4-1                 </span></span>
<span><span class="co">##  [91] vctrs_0.4.1                   rhdf5filters_1.8.0           </span></span>
<span><span class="co">##  [93] pillar_1.7.0                  lifecycle_1.0.1              </span></span>
<span><span class="co">##  [95] BiocManager_1.30.18           jquerylib_0.1.4              </span></span>
<span><span class="co">##  [97] BiocNeighbors_1.14.0          bitops_1.0-7                 </span></span>
<span><span class="co">##  [99] irlba_2.3.5                   httpuv_1.6.5                 </span></span>
<span><span class="co">## [101] rtracklayer_1.56.1            R6_2.5.1                     </span></span>
<span><span class="co">## [103] BiocIO_1.6.0                  promises_1.2.0.1             </span></span>
<span><span class="co">## [105] gridExtra_2.3                 vipor_0.4.5                  </span></span>
<span><span class="co">## [107] codetools_0.2-18              assertthat_0.2.1             </span></span>
<span><span class="co">## [109] rprojroot_2.0.3               rjson_0.2.21                 </span></span>
<span><span class="co">## [111] withr_2.5.0                   GenomicAlignments_1.32.0     </span></span>
<span><span class="co">## [113] Rsamtools_2.12.0              GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">## [115] parallel_4.2.1                hms_1.1.1                    </span></span>
<span><span class="co">## [117] grid_4.2.1                    beachmat_2.12.0              </span></span>
<span><span class="co">## [119] rmarkdown_2.14                DelayedMatrixStats_1.18.0    </span></span>
<span><span class="co">## [121] shiny_1.7.1                   ggbeeswarm_0.6.0             </span></span>
<span><span class="co">## [123] restfulr_0.0.15</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephanie C. Hicks.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
