<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of single-cell analysis in R/Bioconductor • cshlgsd2022</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of single-cell analysis in R/Bioconductor">
<meta property="og:description" content="Overview of single-cell analysis in R/Bioconductor
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cshlgsd2022</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/cshlgsd2022.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro-single-cell-01.html">Introduction to Single-Cell</a>
    </li>
    <li>
      <a href="../articles/intro-single-cell-02.html">Overview of single-cell analysis in R/Bioconductor</a>
    </li>
    <li>
      <a href="../articles/adv-single-cell-01.html">Large-scale clustering in Bioconductor</a>
    </li>
    <li>
      <a href="../articles/intro-spatial-01.html">Introduction to Spatial Transcriptomics Data Science</a>
    </li>
    <li>
      <a href="../articles/adv-spatial-01.html">Advanced Spatial Transcriptomics Data Science</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/index.html">More articles...</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://twitter.com/stephaniehicks" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/stephaniehicks/cshlgsd2022" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of single-cell analysis in
R/Bioconductor</h1>
                        <h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            
      
      
      <div class="hidden name"><code>intro-single-cell-02.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="key-resources">Key resources<a class="anchor" aria-label="anchor" href="#key-resources"></a>
</h3>
<ul>
<li>Workshop material: <a href="https://stephaniehicks.com/cshlgsd2022">pkgdown website</a>
</li>
<li>Code: <a href="https://github.com/stephaniehicks/cshlgsd2022" class="external-link">GitHub</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-1">Part 1<a class="anchor" aria-label="anchor" href="#part-1"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Be able to list at least three quality control (QC) metrics for
scRNA-seq data</li>
<li>Know the difference between fixed and adaptive thresholds for QC
metrics</li>
<li>Be able to define a “library size” in terms of normalization</li>
<li>Describe some common probability distributions used to model
scRNA-seq data</li>
<li>Describe imputation and what it does to scRNA-seq data</li>
</ol>
</div>
<div class="section level3">
<h3 id="materials">Materials<a class="anchor" aria-label="anchor" href="#materials"></a>
</h3>
<p>We will go through the best practices single-cell data science slides
available here:</p>
<ul>
<li><a href="https://docs.google.com/presentation/d/1v-QUVyBefRUlPo_xMwua3jAH_GCP7nAqPQ5wM2KuxVI/edit?usp=sharing" class="external-link uri">https://docs.google.com/presentation/d/1v-QUVyBefRUlPo_xMwua3jAH_GCP7nAqPQ5wM2KuxVI/edit?usp=sharing</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-2">Part 2<a class="anchor" aria-label="anchor" href="#part-2"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives-1">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-1"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Be able to describe a standard workflow for analyzing single-cell
data</li>
<li>Be able to run code for a standard workflow starting from loading a
<code>SingleCellExperiment</code> in R and identifying clusters.</li>
</ol>
</div>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>Here, we provide an overview of the framework of a typical scRNA-seq
analysis workflow:</p>
<p><img src="figures/workflow.png" width="780px" style="display: block; margin: auto;"></p>
<p>In the simplest case, the workflow has the following form:</p>
<ol style="list-style-type: decimal">
<li>We compute quality control metrics to remove low-quality cells that
would interfere with downstream analyses. These cells may have been
damaged during processing or may not have been fully captured by the
sequencing protocol. Common metrics includes the total counts per cell,
the proportion of spike-in or mitochondrial reads and the number of
detected features.</li>
<li>We convert the counts into normalized expression values to eliminate
cell-specific biases (e.g., in capture efficiency). This allows us to
perform explicit comparisons across cells in downstream steps like
clustering. We also apply a transformation, typically log, to adjust for
the mean-variance relationship.</li>
<li>We perform feature selection to pick a subset of interesting
features for downstream analysis. This is done by modelling the variance
across cells for each gene and retaining genes that are highly variable.
The aim is to reduce computational overhead and noise from uninteresting
genes.</li>
<li>We apply dimensionality reduction to compact the data and further
reduce noise. Principal components analysis is typically used to obtain
an initial low-rank representation for more computational work, followed
by more aggressive methods like <span class="math inline">\(t\)</span>-stochastic neighbor embedding for
visualization purposes.</li>
<li>We cluster cells into groups according to similarities in their
(normalized) expression profiles. This aims to obtain groupings that
serve as empirical proxies for distinct biological states. We typically
interpret these groupings by identifying differentially expressed marker
genes between clusters.</li>
</ol>
<p>Today and tomorrow, we will describe each analysis step in more
detail.</p>
</div>
<div class="section level3">
<h3 id="quick-start-simple">Quick start (simple)<a class="anchor" aria-label="anchor" href="#quick-start-simple"></a>
</h3>
<p>Here, we use the a droplet-based retina dataset from Macosko et
al. (2015), provided in the <code>scRNAseq</code> package. This starts
from a count matrix and finishes with clusters in preparation for
biological interpretation. We also demonstrate how to identify
differentially expressed genes between the clusters.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/MacoskoRetinaData.html" class="external-link">MacoskoRetinaData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quality control (using mitochondrial genes).</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/quickPerCellQC.html" class="external-link">quickPerCellQC</a></span><span class="op">(</span><span class="va">qcstats</span>, percent_subsets<span class="op">=</span><span class="st">"subsets_Mito_percent"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">filtered</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Normalization.</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents<span class="op">=</span><span class="fl">25</span>, subset_row<span class="op">=</span><span class="va">hvg</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Clustering.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred<span class="op">=</span><span class="st">'PCA'</span>,</span>
<span>    BLUSPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>cluster.fun<span class="op">=</span><span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span>    </span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualization.</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"label"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="intro-single-cell-02_files/figure-html/unnamed-chunk-3-1.png" alt="UMAP plot of the retina dataset, where each point is a cell and is colored by the assigned cluster identity." width="700"><p class="caption">
UMAP plot of the retina dataset, where each point is a cell and is
colored by the assigned cluster identity.
</p>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Marker detection (between pairs of groups)</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sce</span>, test.type<span class="op">=</span><span class="st">"wilcox"</span>, direction<span class="op">=</span><span class="st">"up"</span>, lfc<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 13</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 24658 rows and 16 columns</span></span>
<span><span class="co">##                     Top      p.value          FDR summary.AUC     AUC.2</span></span>
<span><span class="co">##               &lt;integer&gt;    &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## MEG3                  1  0.00000e+00  0.00000e+00    0.867306  0.857315</span></span>
<span><span class="co">## TUBA1A                1  3.58784e-81  5.89793e-78    0.609195  0.479862</span></span>
<span><span class="co">## SNHG11                1  0.00000e+00  0.00000e+00    0.737419  0.747343</span></span>
<span><span class="co">## SYT1                  2 3.51850e-268 1.08449e-264    0.786614  0.454677</span></span>
<span><span class="co">## CALM1                 2  0.00000e+00  0.00000e+00    0.812019  0.665619</span></span>
<span><span class="co">## ...                 ...          ...          ...         ...       ...</span></span>
<span><span class="co">## VSIG1             24654            1            1           0         0</span></span>
<span><span class="co">## GM16390           24655            1            1           0         0</span></span>
<span><span class="co">## GM25207           24656            1            1           0         0</span></span>
<span><span class="co">## 1110059M19RIK     24657            1            1           0         0</span></span>
<span><span class="co">## GM20861           24658            1            1           0         0</span></span>
<span><span class="co">##                   AUC.3     AUC.4     AUC.5     AUC.6     AUC.7     AUC.8</span></span>
<span><span class="co">##               &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## MEG3          0.6814265  0.463403  0.806255  0.478814 0.1822596  0.888640</span></span>
<span><span class="co">## TUBA1A        0.1162662  0.315768  0.422655  0.289249 0.4529957  0.541010</span></span>
<span><span class="co">## SNHG11        0.0822847  0.562474  0.741261  0.703891 0.0228628  0.742098</span></span>
<span><span class="co">## SYT1          0.5354274  0.317921  0.786614  0.307464 0.0591646  0.549534</span></span>
<span><span class="co">## CALM1         0.1535742  0.370682  0.715928  0.425258 0.0105809  0.791806</span></span>
<span><span class="co">## ...                 ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">## VSIG1                 0         0         0         0         0         0</span></span>
<span><span class="co">## GM16390               0         0         0         0         0         0</span></span>
<span><span class="co">## GM25207               0         0         0         0         0         0</span></span>
<span><span class="co">## 1110059M19RIK         0         0         0         0         0         0</span></span>
<span><span class="co">## GM20861               0         0         0         0         0         0</span></span>
<span><span class="co">##                   AUC.9     AUC.10    AUC.11    AUC.12    AUC.13</span></span>
<span><span class="co">##               &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## MEG3           0.850705 0.48570145  0.866916  0.873796  0.867306</span></span>
<span><span class="co">## TUBA1A         0.509171 0.33506617  0.552997  0.566389  0.609195</span></span>
<span><span class="co">## SNHG11         0.745361 0.74450392  0.737413  0.744927  0.737419</span></span>
<span><span class="co">## SYT1           0.823158 0.39904104  0.575948  0.584176  0.633476</span></span>
<span><span class="co">## CALM1          0.444978 0.00587859  0.749949  0.834002  0.812019</span></span>
<span><span class="co">## ...                 ...        ...       ...       ...       ...</span></span>
<span><span class="co">## VSIG1                 0          0         0         0         0</span></span>
<span><span class="co">## GM16390               0          0         0         0         0</span></span>
<span><span class="co">## GM25207               0          0         0         0         0</span></span>
<span><span class="co">## 1110059M19RIK         0          0         0         0         0</span></span>
<span><span class="co">## GM20861               0          0         0         0         0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="quick-start-multiple-batches">Quick start (multiple batches)<a class="anchor" aria-label="anchor" href="#quick-start-multiple-batches"></a>
</h3>
<p>If you could like to see Quick Start for scRNA-seq data with multiple
batches, check out this workflow:</p>
<ul>
<li><a href="https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-multiple-batches" class="external-link uri">https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-multiple-batches</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin21.5.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /opt/homebrew/Cellar/openblas/0.3.20/lib/libopenblasp-r0.3.20.dylib</span></span>
<span><span class="co">## LAPACK: /opt/homebrew/Cellar/r/4.2.1/lib/R/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] bluster_1.6.0               scran_1.24.0               </span></span>
<span><span class="co">##  [3] scater_1.24.0               ggplot2_3.3.6              </span></span>
<span><span class="co">##  [5] scuttle_1.6.2               scRNAseq_2.10.0            </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.18.0 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">##  [9] Biobase_2.56.0              GenomicRanges_1.48.0       </span></span>
<span><span class="co">## [11] GenomeInfoDb_1.32.2         IRanges_2.30.0             </span></span>
<span><span class="co">## [13] S4Vectors_0.34.0            BiocGenerics_0.42.0        </span></span>
<span><span class="co">## [15] MatrixGenerics_1.8.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] AnnotationHub_3.4.0           BiocFileCache_2.4.0          </span></span>
<span><span class="co">##   [3] systemfonts_1.0.4             igraph_1.3.2                 </span></span>
<span><span class="co">##   [5] lazyeval_0.2.2                BiocParallel_1.30.3          </span></span>
<span><span class="co">##   [7] digest_0.6.29                 ensembldb_2.20.2             </span></span>
<span><span class="co">##   [9] htmltools_0.5.2               viridis_0.6.2                </span></span>
<span><span class="co">##  [11] fansi_1.0.3                   magrittr_2.0.3               </span></span>
<span><span class="co">##  [13] memoise_2.0.1                 ScaledMatrix_1.4.0           </span></span>
<span><span class="co">##  [15] cluster_2.1.3                 limma_3.52.2                 </span></span>
<span><span class="co">##  [17] Biostrings_2.64.0             pkgdown_2.0.5                </span></span>
<span><span class="co">##  [19] prettyunits_1.1.1             colorspace_2.0-3             </span></span>
<span><span class="co">##  [21] blob_1.2.3                    rappdirs_0.3.3               </span></span>
<span><span class="co">##  [23] ggrepel_0.9.1                 textshaping_0.3.6            </span></span>
<span><span class="co">##  [25] xfun_0.31                     dplyr_1.0.9                  </span></span>
<span><span class="co">##  [27] crayon_1.5.1                  RCurl_1.98-1.7               </span></span>
<span><span class="co">##  [29] jsonlite_1.8.0                glue_1.6.2                   </span></span>
<span><span class="co">##  [31] gtable_0.3.0                  zlibbioc_1.42.0              </span></span>
<span><span class="co">##  [33] XVector_0.36.0                DelayedArray_0.22.0          </span></span>
<span><span class="co">##  [35] BiocSingular_1.12.0           scales_1.2.0                 </span></span>
<span><span class="co">##  [37] edgeR_3.38.1                  DBI_1.1.3                    </span></span>
<span><span class="co">##  [39] Rcpp_1.0.8.3                  viridisLite_0.4.0            </span></span>
<span><span class="co">##  [41] xtable_1.8-4                  progress_1.2.2               </span></span>
<span><span class="co">##  [43] dqrng_0.3.0                   bit_4.0.4                    </span></span>
<span><span class="co">##  [45] rsvd_1.0.5                    metapod_1.4.0                </span></span>
<span><span class="co">##  [47] httr_1.4.3                    ellipsis_0.3.2               </span></span>
<span><span class="co">##  [49] farver_2.1.0                  pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [51] XML_3.99-0.10                 uwot_0.1.11                  </span></span>
<span><span class="co">##  [53] sass_0.4.1                    dbplyr_2.2.1                 </span></span>
<span><span class="co">##  [55] locfit_1.5-9.5                utf8_1.2.2                   </span></span>
<span><span class="co">##  [57] labeling_0.4.2                tidyselect_1.1.2             </span></span>
<span><span class="co">##  [59] rlang_1.0.3                   later_1.3.0                  </span></span>
<span><span class="co">##  [61] AnnotationDbi_1.58.0          munsell_0.5.0                </span></span>
<span><span class="co">##  [63] BiocVersion_3.15.2            tools_4.2.1                  </span></span>
<span><span class="co">##  [65] cachem_1.0.6                  cli_3.3.0                    </span></span>
<span><span class="co">##  [67] generics_0.1.2                RSQLite_2.2.14               </span></span>
<span><span class="co">##  [69] ExperimentHub_2.4.0           evaluate_0.15                </span></span>
<span><span class="co">##  [71] stringr_1.4.0                 fastmap_1.1.0                </span></span>
<span><span class="co">##  [73] yaml_2.3.5                    ragg_1.2.2                   </span></span>
<span><span class="co">##  [75] knitr_1.39                    bit64_4.0.5                  </span></span>
<span><span class="co">##  [77] fs_1.5.2                      purrr_0.3.4                  </span></span>
<span><span class="co">##  [79] KEGGREST_1.36.2               AnnotationFilter_1.20.0      </span></span>
<span><span class="co">##  [81] sparseMatrixStats_1.8.0       mime_0.12                    </span></span>
<span><span class="co">##  [83] xml2_1.3.3                    biomaRt_2.52.0               </span></span>
<span><span class="co">##  [85] compiler_4.2.1                rstudioapi_0.13              </span></span>
<span><span class="co">##  [87] beeswarm_0.4.0                filelock_1.0.2               </span></span>
<span><span class="co">##  [89] curl_4.3.2                    png_0.1-7                    </span></span>
<span><span class="co">##  [91] interactiveDisplayBase_1.34.0 statmod_1.4.36               </span></span>
<span><span class="co">##  [93] tibble_3.1.7                  bslib_0.3.1                  </span></span>
<span><span class="co">##  [95] stringi_1.7.6                 highr_0.9                    </span></span>
<span><span class="co">##  [97] RSpectra_0.16-1               GenomicFeatures_1.48.3       </span></span>
<span><span class="co">##  [99] desc_1.4.1                    lattice_0.20-45              </span></span>
<span><span class="co">## [101] ProtGenerics_1.28.0           Matrix_1.4-1                 </span></span>
<span><span class="co">## [103] vctrs_0.4.1                   pillar_1.7.0                 </span></span>
<span><span class="co">## [105] lifecycle_1.0.1               BiocManager_1.30.18          </span></span>
<span><span class="co">## [107] jquerylib_0.1.4               RcppAnnoy_0.0.19             </span></span>
<span><span class="co">## [109] BiocNeighbors_1.14.0          bitops_1.0-7                 </span></span>
<span><span class="co">## [111] irlba_2.3.5                   httpuv_1.6.5                 </span></span>
<span><span class="co">## [113] rtracklayer_1.56.1            R6_2.5.1                     </span></span>
<span><span class="co">## [115] BiocIO_1.6.0                  promises_1.2.0.1             </span></span>
<span><span class="co">## [117] gridExtra_2.3                 vipor_0.4.5                  </span></span>
<span><span class="co">## [119] codetools_0.2-18              assertthat_0.2.1             </span></span>
<span><span class="co">## [121] rprojroot_2.0.3               rjson_0.2.21                 </span></span>
<span><span class="co">## [123] withr_2.5.0                   GenomicAlignments_1.32.0     </span></span>
<span><span class="co">## [125] Rsamtools_2.12.0              GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">## [127] parallel_4.2.1                hms_1.1.1                    </span></span>
<span><span class="co">## [129] grid_4.2.1                    beachmat_2.12.0              </span></span>
<span><span class="co">## [131] rmarkdown_2.14                DelayedMatrixStats_1.18.0    </span></span>
<span><span class="co">## [133] shiny_1.7.1                   ggbeeswarm_0.6.0             </span></span>
<span><span class="co">## [135] restfulr_0.0.15</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephanie C. Hicks.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
